### PR TITLE
Honor channels and an alternate depot URL when exporting packages

### PR DESCRIPTION
**Pull request by reset**
* Exporters will now by default be downloaded from the stable
  channel before running
* Add `--channel` flag to `pkg export`
* Add `--url` flag to `pkg export`

tenor-201683009

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by adamhjk** [comment]
So the actual changes to the studio are already done? This was literally just not piped through?

tenor-196572730

**Comment by reset** [comment]
@adamhjk that's what it looks like 😢

**Comment by mwrock** [comment]
we can ignore the appveyor breaks until https://github.com/habitat-sh/habitat/pull/2842 gets through

**Comment by sdelano** [comment]
@reset does this mean that if I do `hab pkg export ...` that I'll also have to specify `--url` and `--channel`, or will that get picked up by default from what I used to install `hab`?

**Comment by adamhjk** [comment]
tl;dr - this commit makes it so you get the stable channel for the supervisor, but what you actually want is two flags that pipe down through the shell code:

`--channel FOO`, which selects the channel for the package you want exported

`--supervisor-channel FOO`, whihc selects the channel for the supervisor version in the export.

**Comment by adamhjk** [comment]
@sdelano that's a good suggestiont o default to whatever the hab you're running is, tho

**Comment by sdelano** [comment]
@adamhjk if I'm reading the code correctly though the URL and CHANNEL params get passed to the install portion of `hab pkg export`, which actually installs the `hab-pkg-$type` package.

The origination of this PR is that:

> I install `hab` from stable. I run `hab package export`, but don't have the `hab-pkg-dockerizer` package installed, so when hab goes to install it I actually get an unstable version that doesn't correspond to the version of `hab` I have.

**Comment by reset** [comment]
Another commit is going to be pushed up which exposes two channel and two URL flags. This will allow you to specify the toolchain's channel/url and your package's channel/url.

You should very rarely use the one to pull in an unstable toolchain but you _may_ want to use the one for your package itself

**Comment by reset** [comment]
@adamhjk @sdelano we're good to go here

**Comment by reset** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/258380542).

**Comment by reset** [comment]
@fnichol and I discussed overall improvements to some of our studio/exporter logic which will make this easier to maintain. We'll tackle those in another PR

File: components/hab/src/main.rs [hunk 1]
```diff
@@ -324,10 +325,27 @@ fn sub_pkg_exec(m: &ArgMatches, cmd_args: Vec<OsString>) -> Result<()> {
...
+    let format = &m.value_of("FORMAT").unwrap();
+    let env_or_default = henv::var(DEPOT_URL_ENVVAR).unwrap_or(DEFAULT_DEPOT_URL.to_string());
+    let url = m.value_of("DEPOT_URL").unwrap_or(&env_or_default);
+    let channel = m.value_of("CHANNEL")
+        .and_then(|c| Some(c.to_string()))
+        .unwrap_or(channel::default());
+    let hab_url = m.value_of("HAB_DEPOT_URL").unwrap_or(&env_or_default);
+    let hab_channel = m.value_of("HAB_CHANNEL")
+        .and_then(|c| Some(c.to_string()))
+        .unwrap_or(channel::default());
```
Comment (sdelano) [inline components/hab/src/main.rs]: `channel::default()` actually looks like it'll look to `DEPOT_CHANNEL_ENVVAR` first and then eventually `"stable"` [link](https://github.com/habitat-sh/habitat/blob/master/components/core/src/channel.rs#L29). I think that makes sense for `channel` but for `hab_channel` do we want to default to stable here or go through that extra layer of the other envvar?

Comment (reset) [inline components/hab/src/main.rs]: I think it's appropriate for here, yeah. The env vars are just to establish an escape hatch to override the default values. They're largely supposed to be used from within a `.studiorc` for dev purposes.

We could of course create another escape hatch for just this command but I'd rather just suggest people to drive the CLI with the flags

**Comment by thesentinels** [comment]
:broken_heart: Travis CI reports [this PR failed to pass the test suite](https://travis-ci.org/habitat-sh/habitat/builds/258380542).

The next step is to examine the [job](https://travis-ci.org/habitat-sh/habitat/builds/258380542) and figure out why. If it is transient, you can try re-triggering the Travis CI Job - if it passes, this PR will be automatically merged. If it is not transient, you should fix the issue and update this pull request, and issue `approve` again. If you believe it will never pass, and you are feeling :godmode:, you can issue a `force` to merge this PR anyway.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/258380542).

File: components/hab/src/main.rs (hunk 1 above)
Comment (sdelano) [inline components/hab/src/main.rs]: Sounds good to me!

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/258380542).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265
